### PR TITLE
Work to make the DLT LSTM work with float16.

### PR DESCRIPTION
**Pull request by abergeron**

### CODE REVIEW

File: theano/tensor/nnet/sigm.py [hunk 1]
```diff
@@ -59,15 +59,8 @@ def c_code(self, node, name, inp, out, sub):
         # 0. The reset of the logic always generate 0 or 1 in those
         # cases. This is a speed optimization.
         # The constants were obtained by looking at the output of python commands like:
```
Comment (lamblin) [inline theano/tensor/nnet/sigm.py]: Update the comment if you remove the code it is referring to.

**Comment by nouiz** [comment]
In the ticket for f16 follow up, put all the op that you marked as float16 safe and tell we need to test them. Put as an item, make the list of op that isn't supported on CPU/GPU.

File: theano/compile/debugmode.py [hunk 2]
```diff
@@ -1708,10 +1708,20 @@ def make_all(self, profiler=None, input_storage=None
...
+
+                # Don't try to test the C code for float16 if not
+                # tagged ok.
+                if (not getattr(node.op, '_f16_ok', False) and
+                        (any(getattr(i, 'dtype', '') == 'float16'
+                             for i in node.inputs) or
+                         any(getattr(o, 'dtype', '') == 'float16'
+                             for o in node.outputs))):
+                    raise utils.MethodNotDefined()
+
```
Comment (nouiz) [inline theano/compile/debugmode.py]: This continue to add duplicated code here from Op.make_thunk(). Could you refactor that in a method make_c_thunk in the file gof/op.py?

File: theano/tensor/nnet/sigm.py (hunk 1 above)
Comment (abergeron) [inline theano/tensor/nnet/sigm.py]: I will undelete and reformat it.  It looked like dead code.

File: theano/compile/debugmode.py (hunk 2 above)
Comment (abergeron) [inline theano/compile/debugmode.py]: I'll also do it for python thunks.

**Comment by abergeron** [comment]
I've made a series of changes that should address all comments.

**Comment by abergeron** [comment]
I fixed the problem with DebugMode that was the source of failing tests.  I'm not 100% convinced that the change is the right one since I make the python and C thunk share the compute map.

If you have a better solution, I'm all ears.

**Comment by kyunghyuncho** [comment]
@nouiz @abergeron Does it mean that any code can be run with float16 as long as it only includes those OPs used in DLT LSTM tutorial? Does simply setting floatX=float16 work?

**Comment by nouiz** [comment]
Not yet.

**Pull request merged**